File: tests-wasm/wasm.js
const child_process = require('child_process');
const assert = require('assert');
const test = require('node:test');
const wasm = require('../build/tests-wasm/wasm');

function toJS(obj) {
  const result = {};
  for (const key of Object.keys(obj.__proto__)) {
       result[key] = typeof obj[key] === "object" ? toJS(obj[key]) : obj[key];
  }
  return result;
}

const expected = {
  "result": "success",
  "href": "https://google.com/?q=Yagiz#Nizipli",
  "type": 2,
  "components": {
    "protocol_end": 6,
    "username_end": 8,
    "host_start": 8,
    "host_end": 18,
    "port": 4294967295,
    "pathname_start": 18,
    "search_start": 19,
    "hash_start": 27
  }
};

test('wasm', async () => {
  const { parse } = await wasm();
  console.log();
  assert.deepStrictEqual(toJS(parse('https://google.com/?q=Yagiz#Nizipli')), expected);
});

File: tests-wasm/wasm.cpp
#include "ada.h"
#include <emscripten/emscripten.h>
#include <emscripten/bind.h>

using namespace emscripten;

struct parse_result {
  std::string result;
  std::string href;
  uint32_t type;
  ada::url_components components;
};

parse_result parse(const std::string &input) {
  auto out = ada::parse<ada::url_aggregator>(input);
  parse_result result;
  if (!out.has_value()) {
    result.result = "fail";
  } else {
    result.result = "success";
    result.href = std::string(out->get_href());
    result.type = out->type;
    result.components = out->get_components();
  }
  return result;
}

EMSCRIPTEN_BINDINGS(url_components) {
  class_<parse_result>("Result")
      .property("result", &parse_result::result)
      .property("href", &parse_result::href)
      .property("type", &parse_result::type)
      .property("components", &parse_result::components);
  class_<ada::url_components>("URLComponents")
      .property("protocol_end", &ada::url_components::protocol_end)
      .property("username_end", &ada::url_components::username_end)
      .property("host_start", &ada::url_components::host_start)
      .property("host_end", &ada::url_components::host_end)
      .property("port", &ada::url_components::port)
      .property("pathname_start", &ada::url_components::pathname_start)
      .property("search_start", &ada::url_components::search_start)
      .property("hash_start", &ada::url_components::hash_start);

  function("parse", &parse);
}